src/prommis/uky/costing/ree_plant_capcost.py

```diff
@@ -2675,3 +2736,672 @@ def costing_upper_bound_eq(c, p):
+                    "with attributes total_BEC, total_installation_cost, "
+                    "total_fixed_OM_cost, total_variable_OM_cost, "
+                    "other_plant_costs, land_cost, and total_sales_revenue. "
+                    "Please confirm that b is a FlowsheetCostingBlockData object "
+                    "and that all expected attributes exist."
+                )
+        elif isinstance(b, FlowsheetBlock):
+            # if b is not a costing block, it must be a flowsheet block
+            # variables and constraints will be added there
+
+            if None in [
+                total_capital_cost,
+                annual_operating_cost,
+                annual_operating_cost,
+                cost_year,
+            ]:
+                raise AttributeError(
+                    "If b is not passed as a FlowsheetcostingBlockData object, "
+                    "then total_capital_cost, annual_operating_cost, and annual_revenue "
+                    "must be passed, and cost_year must be passed as a string, e.g. '2021'."
+                )
+
+            else:
+                # check if the cost arguments are variables or expressions with units and handle appropriately
+                costs = {
+                    "CAPEX": total_capital_cost,
+                    "OPEX": annual_operating_cost,
+                    "REVENUE": annual_revenue,
+                }
+                cost_units = getattr(pyunits, "MUSD_" + cost_year)
+
+                for key in costs.keys():
+                    if type(key) in [Expression, ScalarExpression]:
+                        if pyunits.get_units(key) == pyunits.dimensionless:
+                            costs[key] = Expression(expr=costs[key].expr * cost_units)
+                        else:
+                            costs[key] = Expression(
+                                expr=pyunits.convert(
+                                    costs[key].expr, to_units=cost_units
+                                )
+                            )
+                    else:
+                        if pyunits.get_units(key) == pyunits.dimensionless:
+                            costs[key] = Expression(expr=costs[key] * cost_units)
+                        else:
+                            costs[key] = Expression(
+                                expr=pyunits.convert(costs[key], to_units=cost_units)
+                            )
+
+                # store for later use
+                CAPEX = costs["CAPEX"]
+                OPEX = costs["OPEX"]
+                REVENUE = costs["REVENUE"]
+        else:
+            # another block type is not allowed
+            raise TypeError(
+                f"Argument {b} is not a valid block type. The argument b "
+                "must be passed a FlowsheetCostingBlockData or FlowsheetBlock object."
+            )
+
+        # check capital expenditure arguments
+        if capital_expenditure_percentages is not None:
```

Comment:
Similarity, with a `ConfigDict` for the arguments, all of this can be broken out into separate verification methods that can be tested in isolation.